lib/puppet/application_support.rb

```diff
@@ -25,10 +26,12 @@ def self.push_application_context(run_mode)
       configured_environment_name = Puppet[:environment]
       if run_mode.name == :agent
         configured_environment = Puppet::Node::Environment.remote(configured_environment_name)
+      elsif environment_mode == :not_required
+        configured_environment = Puppet.lookup(:environments).get(configured_environment_name)
       else
         configured_environment = Puppet.lookup(:environments).get!(configured_environment_name)
       end
-      configured_environment = configured_environment.override_from_commandline(Puppet.settings)
+      configured_environment = configured_environment.override_from_commandline(Puppet.settings) if configured_environment
 
       # Setup a new context using the app's configuration
       Puppet.push_context({:current_environment => configured_environment},
```

Comment:
Is it safe to push a `nil` current environment? I thoughts other parts of the code assume there is always a current environment, even though it may refer a "remote" environment. Wondering if we instead need to push a `Puppet::Node::Environment.remote(configured_environment_name)`? /cc @hlindberg

Reply:
@hlindberg thoughts on ^?

Reply:
While the rest may survive a nil environment by trying to figure out which env to use it may not be the env it should be. So, if it is known that the env is indeed a remote one then use that.

Reply:
So, my interpretation on these comments is that for puppet subcommands like `config` and `help`, we should first try to load a local environment, and if that's not present then load the remote environment. That way the environment would never be nil. Is that correct?

Reply:
👍 to loading the local environment first, because `puppet help` is supposed to show help for built-in applications, e.g. `lib/puppet/applications/agent.rb`, as well as applications installed in modules. And if that fails because the environment doesn't exist locally, then fallback to assuming it's a remote environment.